botocore/client.py

```diff
@@ -106,22 +108,25 @@ def _register_retries(self, service_model):
         self._event_emitter.register('needs-retry.%s' % endpoint_prefix,
                                      handler, unique_id=unique_id)
 
-    def _get_signature_version_and_region(self, service_model, region_name,
-                                          is_secure, scoped_config,
-                                          endpoint_url):
-        # Get endpoint heuristic overrides before creating the
-        # request signer.
-        resolver = self._endpoint_resolver
-        scheme = 'https' if is_secure else 'http'
-        endpoint_config = resolver.construct_endpoint(
-            service_model.endpoint_prefix, region_name, scheme=scheme)
-
-        # Signature version override from endpoint.
+    def _get_preferred_signature_version(self, versions):
+        # Use signers in the following preference order.
+        preference_order = ['s3', 's3v4', 'v4', 'v2']
```

Comment:
Could you explain how this works? I see that the endpoints file have a `signatureVersion` key in their config, for example, S3 has `signatureVersion: ['s3', 's3v4']`.  I don't quite follow why the client creator needs to know about signature order preferences.